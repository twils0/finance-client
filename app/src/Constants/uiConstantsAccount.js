export const actionTypes = {
  SET_CURRENT_FORM: 'ACCOUNT_SET_CURRENT_FORM',
  SET_FORM_HEIGHT: 'ACCOUNT_SET_FORM_HEIGHT',
  SET_FORM_EDIT: 'ACCOUNT_SET_FORM_EDIT',
  SET_INPUT_VALUE_ERROR: 'ACCOUNT_SET_INPUT_VALUE_ERROR',
  SET_BUTTON_TEXT: 'ACCOUNT_SET_BUTTON_TEXT',
  SET_BUTTON_VISIBLE: 'ACCOUNT_SET_BUTTON_VISIBLE',
};

export const formNames = {
  CODE: 'codeForm',
  PROFILE: 'profileForm',
  BILLING: 'billingForm',
  CHANGE_PASSWORD: 'changePasswordForm',
  DELETE_ACCOUNT: 'deleteAccountForm',
};

export const buttonNames = {
  PROFILE: 'profileButton',
  BILLING: 'billingButton',
  CHANGE_PASSWORD: 'changePasswordButton',
  DELETE_ACCOUNT: 'deleteAccountButton',
  FIRST: 'firstButton',
  SECOND: 'secondButton',
};

export const buttonTexts = {
  RESEND: 'Resend',
  SUBMIT: 'Submit',
  EDIT: 'Edit',
  UPDATE: 'Update',
  CANCEL: 'Cancel',
  DELETE: 'Delete',
  CONFIRM: 'Confirm',
};

export const inputNames = {
  [formNames.CODE]: {
    CODE_PHONE: 'codeCodePhone',
  },
  [formNames.PROFILE]: {
    NAME: 'profileName',
    EMAIL: 'profileEmail',
    EMAIL_ADDITIONAL: 'profileEmailAdditional',
    PHONE: 'profilePhone',
  },
  [formNames.BILLING]: {
    NAME_ON_CARD: 'billingNameOnCard',
    PROMO_CODE: 'billingPromoCode',
    STRIPE: 'billingStripe',
  },
  [formNames.CHANGE_PASSWORD]: {
    OLD_PASSWORD: 'changePasswordOldPassword',
    NEW_PASSWORD: 'changePasswordNewPassword',
    NEW_PASSWORD2: 'changePasswordNewPassword2',
  },
  [formNames.DELETE_ACCOUNT]: {
    CONFIRM: 'deleteAccountConfirm',
  },
};
