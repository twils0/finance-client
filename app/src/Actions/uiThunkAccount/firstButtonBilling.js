import { fieldNames } from '../../Constants/dataConstantsAccount';
import { errorMessages } from '../../Constants/uiConstantsApp';
import {
  formNames,
  buttonNames,
  buttonTexts,
  inputNames,
} from '../../Constants/uiConstantsAccount';
import { setField } from '../dataActionsAccount';
import requestUpdateStripeFields from '../dataThunkAccount/requestUpdateStripeFields';
import {
  setFormEdit,
  setButtonText,
  setButtonVisible,
  setInputValueError,
} from '../uiActionsAccount';
import handleCancel from '../uiThunkAccount/handleCancel';

const firstButtonBilling = (payload) => {
  if (!Object.prototype.hasOwnProperty.call(payload, 'stripeElement')) {
    throw new Error(`Please enter a value for the 'stripeElement' key - ${JSON.stringify(payload)}`);
  }
  if (!Object.prototype.hasOwnProperty.call(payload, 'resetStripeElement')) {
    throw new Error(`Please enter a value for the 'resetStripeElement' key - ${JSON.stringify(payload)}`);
  }

  return async (dispatch, getState) => {
    const state = getState();
    const secondButtonVisible = state.ui.internal.account.buttons[buttonNames.SECOND].visible;

    if (!secondButtonVisible) {
      const billingFormEdit = state.ui.internal.account.forms[formNames.BILLING].edit;
      const firstButtonText = state.ui.internal.account.buttons[buttonNames.FIRST].text;
      const secondButtonText = state.ui.internal.account.buttons[buttonNames.SECOND].text;

      if (!billingFormEdit) {
        dispatch(setFormEdit({ id: formNames.BILLING, edit: true }));
      }
      if (firstButtonText !== buttonTexts.UPDATE) {
        dispatch(setButtonText({ id: buttonNames.FIRST, text: buttonTexts.UPDATE }));
      }
      if (secondButtonText !== buttonTexts.CANCEL) {
        dispatch(setButtonText({ id: buttonNames.SECOND, text: buttonTexts.CANCEL }));
      }

      dispatch(setButtonVisible({ id: buttonNames.SECOND, visible: true }));
    } else {
      const { forms } = state.ui.internal.account;
      const { fields } = state.data.account;

      const billingNameOnCard =
        forms[formNames.BILLING].inputs[inputNames[formNames.BILLING].NAME_ON_CARD];
      const billingPromoCode =
        forms[formNames.BILLING].inputs[inputNames[formNames.BILLING].PROMO_CODE];
      const billingStripe = forms[formNames.BILLING].inputs[inputNames[formNames.BILLING].STRIPE];

      if (!billingNameOnCard.value) {
        billingNameOnCard.errorMessage = errorMessages.NO_NAME_ON_CARD;
      }

      if (billingStripe.value && billingStripe.value !== 'complete') {
        billingStripe.errorMessage = billingStripe.value;
      }

      if (
        billingNameOnCard.errorMessage ||
        billingPromoCode.errorMessage ||
        billingStripe.errorMessage
      ) {
        dispatch(setInputValueError(billingNameOnCard));
        dispatch(setInputValueError(billingPromoCode));
        dispatch(setInputValueError(billingStripe));
      } else {
        const stripePayload = {};

        if (fields[fieldNames.PROMO_CODE].value !== billingPromoCode.value) {
          stripePayload.promoCode = billingPromoCode.value;
        }

        if (billingStripe.value === 'complete') {
          stripePayload.name = billingNameOnCard.value;
          stripePayload.element = payload.stripeElement;
        }

        if (Object.keys(payload).length > 0) {
          try {
            await dispatch(requestUpdateStripeFields(stripePayload));

            if (fields[fieldNames.NAME_ON_CARD].value !== billingNameOnCard.value) {
              dispatch(setField({ id: fieldNames.NAME_ON_CARD, value: billingNameOnCard.value }));
            }

            if (fields[fieldNames.PROMO_CODE].value !== billingPromoCode.value) {
              dispatch(setField({ id: fieldNames.PROMO_CODE, value: billingPromoCode.value }));
            }

            payload.resetStripeElement();

            dispatch(handleCancel());
          } catch (error) {
            if (typeof error === 'object' && error && error.raw) {
              billingStripe.value = error.raw.message;
              billingStripe.errorMessage = error.raw.message;

              dispatch(setInputValueError(billingStripe));
            } else if (typeof error === 'object' && error && error.code) {
              switch (error.code) {
                case 'coupon_invalid': {
                  billingPromoCode.errorMessage = errorMessages.PROMO_CODE_INVALID;
                  dispatch(setInputValueError(billingPromoCode));
                  break;
                }
                case 'coupon_expired': {
                  billingPromoCode.errorMessage = errorMessages.PROMO_CODE_EXPIRED;
                  dispatch(setInputValueError(billingPromoCode));
                  break;
                }
                default: {
                  billingStripe.errorMessage = errorMessages.INTERNAL_ERROR;
                  dispatch(setInputValueError(billingStripe));
                  break;
                }
              }
            }
          }
        } else {
          dispatch(handleCancel());
        }
      }
    }

    return null;
  };
};

export default firstButtonBilling;
