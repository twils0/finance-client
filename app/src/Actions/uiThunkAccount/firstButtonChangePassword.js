import equals from 'validator/lib/equals';

import { errorMessages } from '../../Constants/uiConstantsApp';
import { formNames, inputNames } from '../../Constants/uiConstantsAccount';
import requestChangePassword from '../dataThunkAuth/requestChangePassword';
import { setInputValueError } from '../uiActionsAccount';
import handleCancel from '../uiThunkAccount/handleCancel';

const firstButtonChangePassword = () => async (dispatch, getState) => {
  const state = getState();
  const { forms } = state.ui.internal.account;

  const changePasswordOldPassword =
    forms[formNames.CHANGE_PASSWORD].inputs[inputNames[formNames.CHANGE_PASSWORD].OLD_PASSWORD];
  const changePasswordNewPassword =
    forms[formNames.CHANGE_PASSWORD].inputs[inputNames[formNames.CHANGE_PASSWORD].NEW_PASSWORD];
  const changePasswordNewPassword2 =
    forms[formNames.CHANGE_PASSWORD].inputs[inputNames[formNames.CHANGE_PASSWORD].NEW_PASSWORD2];
  const passwordRegex = /^(?=.*[A-Za-z])(?=.*\d)(?=.*[$@$!%*#?&])[A-Za-z\d$@$!%*#?&]{8,}$/;

  if (!changePasswordOldPassword.value) {
    changePasswordOldPassword.errorMessage = errorMessages.NO_PASSWORD;
  }

  if (!changePasswordNewPassword.value) {
    changePasswordNewPassword.errorMessage = errorMessages.NO_NEW_PASSWORD;
  } else if (!passwordRegex.test(changePasswordNewPassword.value)) {
    changePasswordNewPassword.errorMessage = errorMessages.INVALID_PASSWORD;
  }

  if (!equals(changePasswordNewPassword.value, changePasswordNewPassword2.value)) {
    changePasswordNewPassword2.errorMessage = errorMessages.NO_MATCH_PASSWORD;
  }

  if (
    changePasswordOldPassword.errorMessage ||
    changePasswordNewPassword.errorMessage ||
    changePasswordNewPassword2.errorMessage
  ) {
    dispatch(setInputValueError(changePasswordOldPassword));
    dispatch(setInputValueError(changePasswordNewPassword));
    dispatch(setInputValueError(changePasswordNewPassword2));
  } else {
    const changePasswordPayload = {
      oldPassword: changePasswordOldPassword.value,
      newPassword: changePasswordNewPassword.value,
    };

    try {
      await dispatch(requestChangePassword(changePasswordPayload));

      dispatch(handleCancel());
    } catch (error) {
      if (
        error.code === 'NotAuthorizedException' &&
        error.message === 'Incorrect username or password.'
      ) {
        changePasswordOldPassword.errorMessage = errorMessages.PASSWORD_NOT_FOUND;
        dispatch(setInputValueError(changePasswordOldPassword));
      } else {
        changePasswordNewPassword2.errorMessage = errorMessages.INTERNAL_ERROR;
        dispatch(setInputValueError(changePasswordNewPassword2));
      }
    }
  }

  return null;
};

export default firstButtonChangePassword;
