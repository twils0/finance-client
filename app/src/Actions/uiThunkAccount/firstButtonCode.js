import { errorMessages } from '../../Constants/uiConstantsApp';
import { formNames, inputNames } from '../../Constants/uiConstantsAccount';
import requestVerifyFieldConfirm from '../dataThunkAuth/requestVerifyFieldConfirm';
import { setInputValueError, setCurrentForm } from '../uiActionsAccount';
import handleCancel from './handleCancel';

const firstButtonCode = () => async (dispatch, getState) => {
  const state = getState();
  const { forms } = state.ui.internal.account;

  const codeCodePhone = forms[formNames.CODE].inputs[inputNames[formNames.CODE].CODE_PHONE];

  if (!codeCodePhone.value) {
    codeCodePhone.errorMessage = errorMessages.NO_VERIFICATION_CODE_PHONE;
  }

  if (codeCodePhone.errorMessage) {
    dispatch(setInputValueError(codeCodePhone));
  } else {
    try {
      await dispatch(
        requestVerifyFieldConfirm({ field: 'phone_number', code: codeCodePhone.value }),
      );

      dispatch(handleCancel());
      dispatch(setCurrentForm({ current: formNames.PROFILE }));
    } catch ({ field, error }) {
      if (typeof error === 'object' && error && error.code) {
        switch (error.code) {
          case 'CodeMismatchException': {
            codeCodePhone.errorMessage = errorMessages.INVALID_VERIFICATION_CODE;
            dispatch(setInputValueError(codeCodePhone));
            break;
          }
          default: {
            codeCodePhone.errorMessage = errorMessages.INTERNAL_ERROR;
            dispatch(setInputValueError(codeCodePhone));
            break;
          }
        }
      }
    }
  }

  return null;
};

export default firstButtonCode;
