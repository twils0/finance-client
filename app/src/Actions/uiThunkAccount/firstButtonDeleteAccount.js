import { errorMessages } from '../../Constants/uiConstantsApp';
import {
  formNames,
  buttonNames,
  buttonTexts,
  inputNames,
} from '../../Constants/uiConstantsAccount';
import requestDeleteAccount from '../dataThunkAuth/requestDeleteAccount';
import {
  setFormEdit,
  setButtonText,
  setButtonVisible,
  setInputValueError,
} from '../uiActionsAccount';

const firstButtonDeleteAccount = () => async (dispatch, getState) => {
  const state = getState();
  const secondButtonVisible = state.ui.internal.account.buttons[buttonNames.SECOND].visible;

  if (!secondButtonVisible) {
    const deleteAccountFormEdit = state.ui.internal.account.forms[formNames.DELETE_ACCOUNT].edit;
    const firstButtonText = state.ui.internal.account.buttons[buttonNames.FIRST].text;
    const secondButtonText = state.ui.internal.account.buttons[buttonNames.SECOND].text;

    if (!deleteAccountFormEdit) {
      dispatch(setFormEdit({ id: formNames.DELETE_ACCOUNT, edit: true }));
    }
    if (firstButtonText !== buttonTexts.CONFIRM) {
      dispatch(setButtonText({ id: buttonNames.FIRST, text: buttonTexts.CONFIRM }));
    }
    if (secondButtonText !== buttonTexts.CANCEL) {
      dispatch(setButtonText({ id: buttonNames.SECOND, text: buttonTexts.CANCEL }));
    }

    dispatch(setButtonVisible({ id: buttonNames.SECOND, visible: true }));
  } else {
    const { forms } = state.ui.internal.account;

    const deleteAccountConfirm =
      forms[formNames.DELETE_ACCOUNT].inputs[inputNames[formNames.DELETE_ACCOUNT].CONFIRM];

    if (deleteAccountConfirm.value !== 'delete') {
      deleteAccountConfirm.errorMessage = errorMessages.INVALID_DELETE_ACCOUNT;
    }

    if (deleteAccountConfirm.errorMessage) {
      dispatch(setInputValueError(deleteAccountConfirm));
    } else {
      try {
        await dispatch(requestDeleteAccount());

        if (deleteAccountConfirm.value || deleteAccountConfirm.errorMessage) {
          deleteAccountConfirm.value = '';
          deleteAccountConfirm.errorMessage = '';

          dispatch(setInputValueError(deleteAccountConfirm));
        }
      } catch (error) {
        deleteAccountConfirm.value = '';
        deleteAccountConfirm.errorMessage = errorMessages.INTERNAL_ERROR;

        dispatch(setInputValueError(deleteAccountConfirm));
      }
    }
  }

  return null;
};

export default firstButtonDeleteAccount;
