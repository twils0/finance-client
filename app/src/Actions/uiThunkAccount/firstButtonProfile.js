import isEmail from 'validator/lib/isEmail';
import isMobilePhone from 'validator/lib/isMobilePhone';

import { codeTypeNames } from '../../Constants/dataConstantsAuth';
import { fieldNames } from '../../Constants/dataConstantsAccount';
import { errorMessages } from '../../Constants/uiConstantsApp';
import {
  formNames,
  buttonNames,
  buttonTexts,
  inputNames,
} from '../../Constants/uiConstantsAccount';
import { setCodeType } from '../dataActionsAuth';
import { setField } from '../dataActionsAccount';
import requestUpdateStripeFields from '../dataThunkAccount/requestUpdateStripeFields';
import requestUpdateDBFields from '../dataThunkAccount/requestUpdateDBFields';
import {
  setFormEdit,
  setButtonText,
  setButtonVisible,
  setInputValueError,
  setCurrentForm,
} from '../uiActionsAccount';
import handleCancel from './handleCancel';

const firstButtonProfile = () => async (dispatch, getState) => {
  const state = getState();
  const secondButtonVisible = state.ui.internal.account.buttons[buttonNames.SECOND].visible;

  if (!secondButtonVisible) {
    const profileFormEdit = state.ui.internal.account.forms[formNames.PROFILE].edit;
    const firstButtonText = state.ui.internal.account.buttons[buttonNames.FIRST].text;
    const secondButtonText = state.ui.internal.account.buttons[buttonNames.SECOND].text;

    if (!profileFormEdit) {
      dispatch(setFormEdit({ id: formNames.PROFILE, edit: true }));
    }
    if (firstButtonText !== buttonTexts.UPDATE) {
      dispatch(setButtonText({ id: buttonNames.FIRST, text: buttonTexts.UPDATE }));
    }
    if (secondButtonText !== buttonTexts.CANCEL) {
      dispatch(setButtonText({ id: buttonNames.SECOND, text: buttonTexts.CANCEL }));
    }

    dispatch(setButtonVisible({ id: buttonNames.SECOND, visible: true }));
  } else {
    const { forms } = state.ui.internal.account;

    const profileName = forms[formNames.PROFILE].inputs[inputNames[formNames.PROFILE].NAME];
    const profileEmail = forms[formNames.PROFILE].inputs[inputNames[formNames.PROFILE].EMAIL];
    const profileEmailAdditional = forms[formNames.PROFILE].inputs[inputNames[formNames.PROFILE].EMAIL_ADDITIONAL];
    const profilePhone = forms[formNames.PROFILE].inputs[inputNames[formNames.PROFILE].PHONE];

    if (!profileName.value) {
      profileName.errorMessage = errorMessages.NO_NAME;
    }

    if (!profileEmail.value) {
      profileEmail.errorMessage = errorMessages.NO_EMAIL;
    } else if (!isEmail(profileEmail.value)) {
      profileEmail.errorMessage = errorMessages.INVALID_EMAIL;
    }

    if (profileEmailAdditional.value && !isEmail(profileEmailAdditional.value)) {
      profileEmailAdditional.errorMessage = errorMessages.INVALID_EMAIL;
    }

    const sanitizedPhone = profilePhone.value.replace(/[-() ]/g, '');
    if (!sanitizedPhone) {
      profilePhone.errorMessage = errorMessages.NO_PHONE;
    } else if (!isMobilePhone(sanitizedPhone, 'en-US')) {
      profilePhone.errorMessage = errorMessages.INVALID_PHONE;
    }

    if (
      profileName.errorMessage
      || profileEmail.errorMessage
      || profileEmailAdditional.errorMessage
      || profilePhone.errorMessage
    ) {
      dispatch(setInputValueError(profileName));
      dispatch(setInputValueError(profileEmail));
      dispatch(setInputValueError(profileEmailAdditional));
      dispatch(setInputValueError(profilePhone));
    } else {
      const { fields } = state.data.account;
      const E164Phone = `+1${sanitizedPhone}`;

      const payload = {};

      if (fields[fieldNames.NAME].value !== profileName.value) {
        payload.name = profileName.value;
      }
      if (fields[fieldNames.EMAIL].value !== profileEmail.value) {
        dispatch(setCodeType({ id: codeTypeNames.VERIFY_EMAIL, needed: true }));
        payload.email = profileEmail.value;
      }
      if (fields[fieldNames.EMAIL_ADDITIONAL].value !== profileEmailAdditional.value) {
        dispatch(setCodeType({ id: codeTypeNames.VERIFY_EMAIL_ADDITIONAL, needed: true }));
        payload.emailAdditional = profileEmailAdditional.value;
      }
      if (fields[fieldNames.PHONE].value !== E164Phone) {
        dispatch(setCodeType({ id: codeTypeNames.VERIFY_PHONE, needed: true }));
        payload.phone = E164Phone;
      }

      if (Object.keys(payload).length > 0) {
        try {
          if (payload.email) {
            await Promise.all([
              dispatch(requestUpdateStripeFields(payload)),
              dispatch(requestUpdateDBFields(payload)),
            ]);
          } else {
            await dispatch(requestUpdateDBFields(payload));
          }

          if (fields[fieldNames.NAME].value !== profileName.value) {
            dispatch(setField({ id: fieldNames.NAME, value: profileName.value }));
          }
          if (fields[fieldNames.EMAIL].value !== profileEmail.value) {
            dispatch(setField({ id: fieldNames.EMAIL, value: profileEmail.value }));
          }
          if (fields[fieldNames.EMAIL_ADDITIONAL].value !== profileEmailAdditional.value) {
            dispatch(
              setField({
                id: fieldNames.EMAIL_ADDITIONAL,
                value: profileEmailAdditional.value,
              }),
            );
          }
          if (fields[fieldNames.PHONE].value !== E164Phone) {
            dispatch(setField({ id: fieldNames.PHONE, value: E164Phone }));
          }

          dispatch(handleCancel());

          if (payload.phone) {
            dispatch(setCurrentForm({ current: formNames.CODE }));
          }
        } catch (error) {
          switch (error.code) {
            case 'AliasExistsException':
              profileEmail.errorMessage = errorMessages.EMAIL_IN_USE_INTERNAL;

              dispatch(setInputValueError(profileEmail));
              break;
            case 'UsernameExistsException':
              profileEmail.errorMessage = errorMessages.EMAIL_IN_USE_INTERNAL;

              dispatch(setInputValueError(profileEmail));
              break;
            default: {
              profilePhone.errorMessage = errorMessages.INTERNAL_ERROR;

              dispatch(setInputValueError(profilePhone));
              break;
            }
          }
        }
      } else {
        dispatch(handleCancel());
      }
    }
  }

  return null;
};

export default firstButtonProfile;
