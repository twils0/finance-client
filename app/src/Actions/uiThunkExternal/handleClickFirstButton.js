import { formNames } from '../../Constants/uiConstantsExternal';
import { setCurrentForm } from '../uiActionsExternal';

import firstButtonResetPassword from './firstButtonResetPassword';
import firstButtonForgotPassword from './firstButtonForgotPassword';
import firstButtonLogin from './firstButtonLogin';
import firstButtonSignUp from './firstButtonSignUp';
import firstButtonCodeMFAPhone from './firstButtonCodeMFAPhone';
import firstButtonCodeVerifyPhone from './firstButtonCodeVerifyPhone';
import firstButtonCodeVerifyEmail from './firstButtonCodeVerifyEmail';
import firstButtonDevice from './firstButtonDevice';

const handleClickFirstButton = (payload) => {
  if (!Object.prototype.hasOwnProperty.call(payload, 'history')) {
    throw new Error(`Please enter a value for the 'history' key - ${JSON.stringify(payload)}`);
  }
  if (!Object.prototype.hasOwnProperty.call(payload, 'match')) {
    throw new Error(`Please enter a value for the 'match' key - ${JSON.stringify(payload)}`);
  }
  if (!Object.prototype.hasOwnProperty.call(payload, 'stripeElement')) {
    throw new Error(
      `Please enter a value for the 'stripeElement' key - ${JSON.stringify(payload)}`,
    );
  }
  if (!Object.prototype.hasOwnProperty.call(payload, 'resetStripeElement')) {
    throw new Error(
      `Please enter a value for the 'resetStripeElement' key - ${JSON.stringify(payload)}`,
    );
  }

  return (dispatch, getState) => {
    const state = getState();
    const { forms } = state.ui.external;

    switch (forms.current) {
      case formNames.RESET_PASSWORD:
        dispatch(firstButtonResetPassword());
        break;
      case formNames.FORGOT_PASSWORD: {
        const { history } = payload;

        dispatch(firstButtonForgotPassword({ history }));
        break;
      }
      case formNames.LOGIN: {
        const { history } = payload;

        dispatch(firstButtonLogin({ history }));
        break;
      }
      case formNames.CARD:
        dispatch(setCurrentForm({ current: formNames.INFO }));
        break;
      case formNames.INFO:
        dispatch(setCurrentForm({ current: formNames.SIGN_UP }));
        break;
      case formNames.SIGN_UP: {
        const { history, stripeElement, resetStripeElement } = payload;

        dispatch(firstButtonSignUp({ history, stripeElement, resetStripeElement }));
        break;
      }
      case formNames.CODE_MFA_PHONE: {
        dispatch(firstButtonCodeMFAPhone());
        break;
      }
      case formNames.CODE_VERIFY_PHONE: {
        const { history } = payload;

        dispatch(firstButtonCodeVerifyPhone({ history }));
        break;
      }
      case formNames.CODE_VERIFY_EMAIL: {
        const { history, match } = payload;

        dispatch(firstButtonCodeVerifyEmail({ history, match }));
        break;
      }
      case formNames.DEVICE: {
        const { history } = payload;

        dispatch(firstButtonDevice({ history }));
        break;
      }
      default:
        break;
    }

    return null;
  };
};

export default handleClickFirstButton;
