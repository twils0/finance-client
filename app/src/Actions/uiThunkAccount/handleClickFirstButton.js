import { formNames } from '../../Constants/uiConstantsAccount';

import firstButtonCode from './firstButtonCode';
import firstButtonProfile from './firstButtonProfile';
import firstButtonBilling from './firstButtonBilling';
import firstButtonChangePassword from './firstButtonChangePassword';
import firstButtonDeleteAccount from './firstButtonDeleteAccount';

const handleClickFirstButton = (payload) => {
  if (!Object.prototype.hasOwnProperty.call(payload, 'stripeElement')) {
    throw new Error(`Please enter a value for the 'stripeElement' key - ${JSON.stringify(payload)}`);
  }
  if (!Object.prototype.hasOwnProperty.call(payload, 'resetStripeElement')) {
    throw new Error(`Please enter a value for the 'resetStripeElement' key - ${JSON.stringify(payload)}`);
  }

  return (dispatch, getState) => {
    const state = getState();
    const { forms } = state.ui.internal.account;

    switch (forms.current) {
      case formNames.CODE:
        dispatch(firstButtonCode());
        break;
      case formNames.PROFILE:
        dispatch(firstButtonProfile());
        break;
      case formNames.BILLING: {
        const { stripeElement, resetStripeElement } = payload;

        dispatch(firstButtonBilling({ stripeElement, resetStripeElement }));
        break;
      }
      case formNames.CHANGE_PASSWORD:
        dispatch(firstButtonChangePassword());
        break;
      case formNames.DELETE_ACCOUNT: {
        dispatch(firstButtonDeleteAccount());
        break;
      }
      default:
        break;
    }

    return null;
  };
};

export default handleClickFirstButton;
