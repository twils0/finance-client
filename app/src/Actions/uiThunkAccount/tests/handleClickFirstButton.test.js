import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { initialState as initialStateUIAccount } from '../../../Reducers/uiReducersAccount';
import { formNames } from '../../../Constants/uiConstantsAccount';
import handleClickFirstButton from '../handleClickFirstButton';
import firstButtonCode from '../firstButtonCode';
import firstButtonProfile from '../firstButtonProfile';
import firstButtonBilling from '../firstButtonBilling';
import firstButtonChangePassword from '../firstButtonChangePassword';
import firstButtonDeleteAccount from '../firstButtonDeleteAccount';

const middleware = [thunk];
const mockStore = configureMockStore(middleware);

jest.mock('../firstButtonCode', () => jest.fn());
jest.mock('../firstButtonProfile', () => jest.fn());
jest.mock('../firstButtonBilling', () => jest.fn());
jest.mock('../firstButtonChangePassword', () => jest.fn());
jest.mock('../firstButtonDeleteAccount', () => jest.fn());

const stripeElement = { test: 'testStripeElement' };
const handlerPayload = {
  stripeElement,
  resetStripeElement: jest.fn(),
};

describe('uiThunkAccount', () => {
  afterEach(() => {
    firstButtonCode.mockReset();
    firstButtonProfile.mockReset();
    firstButtonBilling.mockReset();
    firstButtonChangePassword.mockReset();
    firstButtonDeleteAccount.mockReset();
  });

  describe('handleClickFirstButton', () => {
    it("fails and throws an error when missing 'stripeElement' key in payload", async () => {
      const store = mockStore({
        ui: { internal: { account: initialStateUIAccount } },
      });

      const emptyPayload = {};

      try {
        await store.dispatch(handleClickFirstButton(emptyPayload));
      } catch (errorCatch) {
        expect(errorCatch).toMatchSnapshot();
      }
    });

    it("fails and throws an error when missing 'resetStripeElement' key in payload", async () => {
      const store = mockStore({
        ui: { internal: { account: initialStateUIAccount } },
      });

      const wrongPayload = { stripeElement };

      try {
        await store.dispatch(handleClickFirstButton(wrongPayload));
      } catch (errorCatch) {
        expect(errorCatch).toMatchSnapshot();
      }
    });

    it('calls nothing when forms current is wrong', async () => {
      const stateBeforeUIAccount = JSON.parse(JSON.stringify(initialStateUIAccount));
      stateBeforeUIAccount.forms.current = 'testForm';

      const store = mockStore({ ui: { internal: { account: stateBeforeUIAccount } } });

      const result = store.dispatch(handleClickFirstButton(handlerPayload));

      expect(result).toEqual(null);
    });

    it('calls firstButtonCode when form current is CODE', async () => {
      const stateBeforeUIAccount = JSON.parse(JSON.stringify(initialStateUIAccount));
      stateBeforeUIAccount.forms.current = formNames.CODE;

      const store = mockStore({ ui: { internal: { account: stateBeforeUIAccount } } });

      firstButtonCode.mockReturnValue(() => null);

      const result = store.dispatch(handleClickFirstButton(handlerPayload));

      expect(result).toEqual(null);
      expect(firstButtonCode).toBeCalled();
    });

    it('calls firstButtonProfile when form current is PROFILE', async () => {
      const stateBeforeUIAccount = JSON.parse(JSON.stringify(initialStateUIAccount));
      stateBeforeUIAccount.forms.current = formNames.PROFILE;

      const store = mockStore({ ui: { internal: { account: stateBeforeUIAccount } } });

      firstButtonProfile.mockReturnValue(() => null);

      const result = store.dispatch(handleClickFirstButton(handlerPayload));

      expect(result).toEqual(null);
      expect(firstButtonProfile).toBeCalled();
    });

    it('calls firstButtonBilling when form current is BILLING', async () => {
      const stateBeforeUIAccount = JSON.parse(JSON.stringify(initialStateUIAccount));
      stateBeforeUIAccount.forms.current = formNames.BILLING;

      const store = mockStore({ ui: { internal: { account: stateBeforeUIAccount } } });

      firstButtonBilling.mockReturnValue(() => null);

      const result = store.dispatch(handleClickFirstButton(handlerPayload));

      expect(result).toEqual(null);
      expect(firstButtonBilling).toBeCalledWith(handlerPayload);
    });

    it('calls firstButtonChangePassword when form current is CHANGE_PASSWORD', async () => {
      const stateBeforeUIAccount = JSON.parse(JSON.stringify(initialStateUIAccount));
      stateBeforeUIAccount.forms.current = formNames.CHANGE_PASSWORD;

      const store = mockStore({ ui: { internal: { account: stateBeforeUIAccount } } });

      firstButtonChangePassword.mockReturnValue(() => null);

      const result = store.dispatch(handleClickFirstButton(handlerPayload));

      expect(result).toEqual(null);
      expect(firstButtonChangePassword).toBeCalled();
    });

    it('calls firstButtonDeleteAccount when form current is DELETE_ACCOUNT', async () => {
      const stateBeforeUIAccount = JSON.parse(JSON.stringify(initialStateUIAccount));
      stateBeforeUIAccount.forms.current = formNames.DELETE_ACCOUNT;

      const store = mockStore({ ui: { internal: { account: stateBeforeUIAccount } } });

      firstButtonDeleteAccount.mockReturnValue(() => null);

      const result = store.dispatch(handleClickFirstButton(handlerPayload));

      expect(result).toEqual(null);
      expect(firstButtonDeleteAccount).toBeCalled();
    });
  });
});
