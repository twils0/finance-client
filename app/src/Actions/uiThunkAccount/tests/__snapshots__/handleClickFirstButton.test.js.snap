// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uiThunkAccount handleClickFirstButton fails and throws an error when missing 'resetStripeElement' key in payload 1`] = `[Error: Please enter a value for the 'resetStripeElement' key - {"stripeElement":{"test":"testStripeElement"}}]`;

exports[`uiThunkAccount handleClickFirstButton fails and throws an error when missing 'stripeElement' key in payload 1`] = `[Error: Please enter a value for the 'stripeElement' key - {}]`;
